client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -93,24 +93,32 @@
         private long eventNumber = 0;
         private final ReusableLatch connectionSetup = new ReusableLatch();
         @GuardedBy("lock")
-        private CompletableFuture<Void> emptyInflightFuture = null;
+        private CompletableFuture<Exception> emptyInflightFuture = null;
 
         /**
          * Returns a future that will complete successfully once all the inflight events are acked
          * and will fail if that is not possible for some reason. IE: because the connection
          * dropped, or the segment was sealed.
          */
-        private CompletableFuture<Void> getEmptyInflightFuture() {
+        private CompletableFuture<Exception> getEmptyInflightFuture() {
```

Comment:
I see (towards the end of this file) that you use this to extract the failure exception and then rethrow it. Well, if you simply `completeExceptionally`, and then you verify `isCompletedExceptionally`, simply call `join` and your exception will be thrown (wrapped in a `CompletionException`)